File: app/src/main/java/com/example/kirchhoff/example/ui/recycler/diffutil/ex2/DiffUtilCallback.java
package com.example.kirchhoff.example.ui.recycler.diffutil.ex2;

import android.support.v7.util.DiffUtil;

import java.util.List;

/**
 * @author Kirchhoff-
 */

public class DiffUtilCallback extends DiffUtil.Callback {

    private List<Thing> current;
    private List<Thing> next;

    public DiffUtilCallback(List<Thing> current, List<Thing> next) {
        this.current = current;
        this.next = next;
    }

    @Override
    public int getOldListSize() {
        return current.size();
    }

    @Override
    public int getNewListSize() {
        return next.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Thing currentItem = current.get(oldItemPosition);
        Thing nextItem = next.get(newItemPosition);
        return currentItem.getId() == nextItem.getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Thing currentItem = current.get(oldItemPosition);
        Thing nextItem = next.get(newItemPosition);
        return currentItem.equals(nextItem);
    }
}

File: app/src/main/java/com/example/kirchhoff/example/ui/recycler/diffutil/ex2/DiffUtilCallback.kt
package com.example.kirchhoff.example.ui.recycler.diffutil.ex2

import android.support.v7.util.DiffUtil

class DiffUtilCallback(private val current: List<Thing>,
                       private val next: List<Thing>) : DiffUtil.Callback() {

    override fun getOldListSize() = current.size

    override fun getNewListSize() = next.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) = current[oldItemPosition] == next[newItemPosition]

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) = current[oldItemPosition] == next[newItemPosition]
}
